### ISSUE TITLE
Auto login SSO does not work ?

### ISSUE THREAD

**Issue by fredziak** [oidc-wp/openid-connect-generic#70]
Hi,

I don't know if it is a bug report, a feature request or if I don't understand how to use it but it seems to me that the "auto login SSO" mode is not working, at least not as I expected it to work.

I installed and configured the wordpress plugin in order to communicate with a Keycloak OpenID provider.
I chose the "auto login SSO" because the provider is already used by other applications but in the case of a user having already logged in (from another application), my wordpress site does not detect there is already an open session.

If I check the "Enforce Privacy" option, the previous case works : user is automatically logged in as expected (thus communication with the provider is ok). However, I cannot keep this option because the main parts of my wordpress site should remain public, most content should be readable without authentication.

In other words, I need automatic authentication check (without triggering login on the wordpress site). If the user needs to login twice, this is not SSO and the OpenID provider seems useless.

This plugin is great (thanks for your work) and I'd like to use it. I tried to modify the code to achieve this goal and maybe post a contribution but without success, probably because my knowledge of php, wordpress and openID is limited.
Could you please help me to implement this case ? Thanks in advance.

Frédéric

**Comment by daggerhart** [comment]
Hi Frédéric,

I'm not super familiar with keycloak so you might have to help me out here. When you enable SSO in keycloak does it give you a field where you can list URLs that are part of the SSO group? If so, then I think we can handle this without much work.

I've been searching online for an example of using keycloak setup to manage SSO for its identity providers but haven't found anything. If you could provide me a resource (screenshots or documentation) for how this is done in keycloak that would be great.

**Comment by fredziak** [comment]
Hi,

Thanks a lot for your reply.

The faster way to run keycloak is probably to use the docker image (with embedded database) : https://hub.docker.com/r/jboss/keycloak/ 
Allowed URLs can be defined in admin console at the "client" level : http://www.keycloak.org/docs/latest/server_admin/index.html#_clients

I have allowed the URL corresponding to my wordpress site in the "Valid Redirect URIs" and "Web Origins" fields and I think it is OK as the authentication process succeeds when I activate "Enforce privacy". 
My problem is rather that no authentication process is launched automatically when I go to a a wordpress page without this option (there is absolutely no interaction with the provider). It seems like clicking on "login" is required but as the user has already logged in, the user shoudn't have to do any action at this stage to get an open session in wordpress.

My feeling is that something is missing in the code to achieve such a case, probably not much but this does not mean it is easy...

**Comment by daggerhart** [comment]
Thanks, that's very helpful!

I think the issue boils down to needing something like: "When a user logs into keycloak, keycloak should immediately send them through the authentication process for various service providers". 

This would allow them to be automatically authenticated with WordPress, while users that have not logged into keycloak can still visit the site as normal. I believe this is the "SSO Network" concept that I don't have much experience with it.

If we could figure out how to have keycloak redirect the user through a series of URLs to automatically authenticate them with the various service providers, then we could easily all the WP login page as one of those URLs (with the SSO setting enabled on this plugin). 

Without keycloak being setup to authenticate the user with WP when they login to keycloak, I can't think of an easy way to have this plugin know if a user should be authenticated or not when visiting the WP site.

The reason it's working with the privacy & SSO settings enabled is because the user automatically being redirected to the IDP (keycloak) and is being returned to WP authenticated.

Any thoughts? Am I missing something about what this plugin could be doing to detect the WP visitor was authenticated with keycloak?

**Comment by fredziak** [comment]
Well... I am not a SSO specialist either, but I think it is not the responsibility of the identity provider to notify the applications of login but rather the contrary: every application needs to ask keycloak of the login state. Imagine a quite extreme case where there are 10 or 20 applications, it would be useless to open a session on all of them if the user access only 2 or 3... And if you multiply this by a high number of users, that could create an unnecessary overload of sessions. On top of that, some applications may not be reachable at that time (case of a pure javascript or mobile application). At least, this is the theoretical vision...

Think of the way it works with the Facebook provider for instance. It is when you access your application that it checks if a Facebook session is open and if it is the case authentication is done automatically.

So, if we come back to our case, my feeling is that the plugin in "auto SSO" mode should connect to the provider every time a wordpress page is requested and the user is not logged in from wordpress point of view: if the provider answers that there is no session, the page is displayed anonymously, otherwise it is time to open the wordpress session with the authenticated user.

Of course, there is also the logout case. Should we check the state on the provider every time the user loads a page, which seems a lot of queries, or maybe rely on expiration time information ?

To sum up, I am pretty sure it should work this way but I have no idea how this should be really implemented (to be secure and efficient). Seems like we need advice from an OAuth/OpenID specialist...

**Comment by daggerhart** [comment]
I think you're right. I'll look into some approaches to implement this and ask around.

**Comment by fredziak** [comment]
OK. Thanks. Perhaps it is not the only way to cope with OpenID SSO but I see applications that follow this case, i.e. implicit login when a session is already open in the provider, and it makes sense when you use one. I think the main question as far as the implementation is concerned is how the login state can be efficiently checked.

**Comment by timnolte** [comment]
No longer valid and stale. Closing.

**Comment by HeDo88TH** [comment]
How was this fixed or implemented?

**Comment by timnolte** [comment]
@HeDo88TH there wasn't any "fix" needed. The expectation described wasn't valid. Also, Auto-SSO does work without Privacy Mode turned on as I actively use it on many client sites right now.

**Comment by HeDo88TH** [comment]
> @HeDo88TH there wasn't any "fix" needed. The expectation described wasn't valid. Also, Auto-SSO does work without Privacy Mode turned on as I actively use it on many client sites right now.

Thanks for the quick response. I am using your plugin with KeyCloak and AutoLogin enabled. My user is already authenticated in another app but when going to the private area of my website, It asks me to login again anyways.

I think I am missing something here. I have a private area (every other page except the home) that needs the user to be authenticated. I do this in the `functions.php` file:

```
add_action('template_redirect', 'site_redirect');
function site_redirect()
{
    if (is_user_logged_in()) {
        [...]  
    } else {
        
        if (!is_front_page()) {
            $redirect_auth = do_shortcode('[openid_connect_generic_auth_url]');
	    exit(wp_redirect($redirect_auth));			
        }
    } 
}
```

It redirects the user to the login page but it is not able to understand that the user is already logged in and redirect him back to the requested page. 

Do you think this is a bug?

**Comment by timnolte** [comment]
@HeDo88TH so this isn't so much a bug verses a different use case, or how you are trying to implement this.

I also have 1 client site that has both features turned on, however that means it's for the entire site, and so navigating to the site in any way is sending the user to the IDP. If they are already logged into the IDP then they are sent back to the site and automatically logged into WordPress. If they are already logged into the IDP they are prompted to login there. In this case the IDP is Azure AD, so I'm not sure if this is specifically a Keycloak problem. I have a Keycloak testing instance running so I could test this out.

From what you've posted you aren't using the Privacy feature of the plugin. That is a different use case. I'm not clear though on your statement:
> My user is already authenticated in another app but when going to the private area of my website, It asks me to login again anyways.
If WordPress is asking you to login again then either your hook is not firing at the right time in order to provide the redirect or you are missing an error response from the plugin that is failing the authentication. I will note however that there is a session timeout at play in all of this as well. That means that the IDP may expire the logged in session perhaps within 1 hour, shorter than the WordPress session timeout, and as a result a user may be forced to be logged out of WordPress and then be required to manually click the login button. This can also present itself in a way that you think "I just visited here the first time and it should have automatically logged me in", however, if you had previously logged in your IDP session timeout may have been reached and WordPress forced logged you out. I see this on the client sites I have the Auto-SSO implemented on. I'll come back to a site that I had been logged into and I'm sent to the Login page with just the SSO login button because my IDP session had timed out regardless if the WordPress session had timed out yet.

**Comment by HeDo88TH** [comment]
I am not the admin of KeyCloak, so it is possible that the issue is there. But I expect that the OIP behaves like WordPress: if the user is already logged in and he goes to `wp-login.php`, he gets redirected to the destination page without showing the login form. KeyCloak is not behaving in the same way. If an authenticated user goes to its login page (created with `openid_connect_generic_auth_url`) he does not get redirected to the destination page and has to fill in the login form again.

Is there a way, maybe using some client side javascript, to accomplish this? If it is the inteded behaviour of this plugin I do not know how to implement it then.
